File: agent_serving/serving/rerank/service_reranker.py
"""LLMServiceReranker — rerank via shared llm_service model endpoint."""
from __future__ import annotations

import logging

from agent_serving.serving.schemas.models import (
    QueryPlan,
    QueryUnderstanding,
    RetrievalCandidate,
)

logger = logging.getLogger(__name__)


class LLMServiceReranker:
    """Shared model reranker backed by llm_service /api/v1/models/rerank."""

    # llm_service RerankRequest.documents max_length=200
    _MAX_RERANK_DOCS = 200
    # Max chars per document sent to rerank API
    _MAX_DOC_CHARS = 1000

    def __init__(
        self,
        llm_client,
        *,
        model: str = "rerank",
        top_n: int = 100,
    ) -> None:
        self._llm_client = llm_client
        self._model = model
        self._top_n = top_n

    async def rerank(
        self,
        candidates: list[RetrievalCandidate],
        plan: QueryPlan,
        understanding: QueryUnderstanding | None = None,
    ) -> list[RetrievalCandidate] | None:
        if not candidates or self._llm_client is None:
            return None

        query = understanding.original_query if understanding else " ".join(plan.keywords)
        if not query:
            return None

        top_n = min(self._top_n, len(candidates))
        working_set = candidates[:top_n]

        # Batch rerank to respect API limit
        result = await self._rerank_batched(query, working_set)
        return result

    async def _rerank_batched(
        self,
        query: str,
        candidates: list[RetrievalCandidate],
    ) -> list[RetrievalCandidate] | None:
        """Rerank in batches if candidates exceed API limit, then merge by score."""
        if len(candidates) <= self._MAX_RERANK_DOCS:
            return await self._rerank_single(query, candidates)

        # Split into batches and rerank each
        all_scored: list[tuple[float, RetrievalCandidate]] = []
        for i in range(0, len(candidates), self._MAX_RERANK_DOCS):
            batch = candidates[i:i + self._MAX_RERANK_DOCS]
            batch_result = await self._rerank_single(query, batch)
            if batch_result is None:
                return None
            for c in batch_result:
                all_scored.append((c.score, c))

        # Sort all by score descending
        all_scored.sort(key=lambda x: x[0], reverse=True)
        return [c for _, c in all_scored]

    async def _rerank_single(
        self,
        query: str,
        candidates: list[RetrievalCandidate],
    ) -> list[RetrievalCandidate] | None:
        """Single rerank API call."""
        documents = []
        for candidate in candidates:
            text = candidate.metadata.get("text", "")
            title = candidate.metadata.get("title", "")
            documents.append(f"{title}: {text}" if title else text)

        try:
            response = await self._llm_client.rerank(
                query=query,
                documents=[doc[:self._MAX_DOC_CHARS] for doc in documents],
                model=self._model,
                top_n=len(documents),
            )
        except Exception:
            logger.warning("LLM service rerank call failed", exc_info=True)
            return None

        if not response or not response.get("results"):
            return None

        # Only return candidates that received rerank scores, sorted by relevance
        reordered: list[RetrievalCandidate] = []
        seen_indices: set[int] = set()
        for item in response["results"]:
            idx = item.get("index", -1)
            if 0 <= idx < len(candidates) and idx not in seen_indices:
                seen_indices.add(idx)
                candidate = candidates[idx]
                score = item.get("relevance_score", candidate.score)
                reordered.append(candidate.model_copy(update={"score": float(score)}))

        return reordered

File: agent_serving_java/src/main/java/com/coremasterkb/serving/rerank/ServiceReranker.java
package com.coremasterkb.serving.rerank;

import com.coremasterkb.serving.domain.QueryUnderstanding;
import com.coremasterkb.serving.domain.RetrievalCandidate;
import com.coremasterkb.serving.infrastructure.LlmClient;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.*;

import java.util.*;

/**
 * Reranker backed by the shared LLM service rerank endpoint.
 *
 * <p>Calls the LLM service's dedicated rerank API.
 * Model is managed by llm_service — callers only provide query, documents, and top_n.
 * Returns {@code null} on any failure to signal fallback.
 */
public class ServiceReranker implements Reranker {

    private static final Logger log = LoggerFactory.getLogger(ServiceReranker.class);
    private static final int MAX_RERANK_DOCS = 200;
    private static final int MAX_DOC_CHARS = 1000;
    private static final int DEFAULT_TOP_N = 100;

    private final LlmClient llmClient;
    private final int topN;

    public ServiceReranker(LlmClient llmClient) {
        this(llmClient, DEFAULT_TOP_N);
    }

    public ServiceReranker(LlmClient llmClient, int topN) {
        this.llmClient = llmClient;
        this.topN = topN;
    }

    @Override
    public List<RetrievalCandidate> rerank(List<RetrievalCandidate> candidates, QueryUnderstanding understanding) {
        if (candidates == null || candidates.isEmpty()) {
            return null;
        }
        if (llmClient == null || !llmClient.isAvailable()) {
            return null;
        }

        String query = resolveQuery(understanding);
        if (query == null || query.isBlank()) {
            return null;
        }

        int workingSize = Math.min(topN, candidates.size());
        List<RetrievalCandidate> workingSet = candidates.subList(0, workingSize);

        try {
            if (workingSize <= MAX_RERANK_DOCS) {
                return rerankSingle(query, workingSet);
            }
            return rerankBatched(query, workingSet);
        } catch (Exception e) {
            log.warn("Service rerank call failed: {}", e.getMessage());
            return null;
        }
    }

    private String resolveQuery(QueryUnderstanding understanding) {
        if (understanding != null && understanding.originalQuery() != null
                && !understanding.originalQuery().isBlank()) {
            return understanding.originalQuery();
        }
        return null;
    }

    private List<RetrievalCandidate> rerankBatched(String query, List<RetrievalCandidate> candidates) {
        List<Map.Entry<Double, RetrievalCandidate>> scored = new ArrayList<>();

        for (int i = 0; i < candidates.size(); i += MAX_RERANK_DOCS) {
            int end = Math.min(i + MAX_RERANK_DOCS, candidates.size());
            List<RetrievalCandidate> batch = candidates.subList(i, end);
            List<RetrievalCandidate> batchResult = rerankSingle(query, batch);
            if (batchResult == null) {
                return null;
            }
            for (RetrievalCandidate c : batchResult) {
                scored.add(Map.entry(c.score(), c));
            }
        }

        scored.sort(Map.Entry.<Double, RetrievalCandidate>comparingByKey().reversed());
        return scored.stream().map(Map.Entry::getValue).toList();
    }

    @SuppressWarnings("unchecked")
    private List<RetrievalCandidate> rerankSingle(String query, List<RetrievalCandidate> candidates) {
        List<String> documents = new ArrayList<>();
        for (RetrievalCandidate c : candidates) {
            String text = stringFromMetadata(c, "text");
            String title = stringFromMetadata(c, "title");
            String doc = (title != null && !title.isEmpty())
                    ? title + ": " + text
                    : text;
            documents.add(truncate(doc, MAX_DOC_CHARS));
        }

        // Use LlmClient execute with rerank-specific payload
        Map<String, Object> input = new HashMap<>();
        input.put("query", query);
        input.put("documents", documents);
        input.put("top_n", documents.size());

        Map<String, Object> response = llmClient.execute("reranker", "service-rerank", input);
        if (response == null) {
            return null;
        }

        Object resultsObj = response.get("results");
        if (!(resultsObj instanceof List<?> rawList)) {
            return null;
        }

        Set<Integer> seenIndices = new HashSet<>();
        List<RetrievalCandidate> reordered = new ArrayList<>();

        for (Object item : rawList) {
            if (!(item instanceof Map<?, ?> m)) continue;
            Map<String, Object> r = (Map<String, Object>) m;

            Object idxObj = r.get("index");
            Object scoreObj = r.get("relevance_score");
            if (!(idxObj instanceof Number) || !(scoreObj instanceof Number)) continue;

            int idx = ((Number) idxObj).intValue();
            if (idx < 0 || idx >= candidates.size() || seenIndices.contains(idx)) continue;

            seenIndices.add(idx);
            double score = ((Number) scoreObj).doubleValue();
            reordered.add(candidates.get(idx).withScore(score));
        }

        return reordered.isEmpty() ? null : reordered;
    }

    private static String stringFromMetadata(RetrievalCandidate c, String key) {
        Object val = c.metadata().get(key);
        return val instanceof String s ? s : "";
    }

    private static String truncate(String s, int maxLen) {
        if (s == null) return "";
        return s.length() <= maxLen ? s : s.substring(0, maxLen);
    }
}
